services/queued/internal/queue/queue.go

```diff
@@ -228,249 +234,279 @@ func zipOutputToBase64(output_path string) (string, error) {
 
-				for file, content := range msg.Files {
-					path := filepath.Join(submissionVolume.HostPath, file)
+				if resp.StatusCode == 404 {
+					log.Println("Empty response. No messages in input queue.")
+					isOk = false
 
-					err := ioutil.WriteFile(path, []byte(content), 0644)
-					if err != nil {
-						log.Println("<3>Cannot write submission file", err)
-						return err.Error()
+					// check for idle time
+					diff := time.Since(lastGoodRead).Seconds()
+					log.Println("Idle time: ", diff, " s.")
+
+					if diff > arguments.MaxIdleTime {
+
+						log.Println("max idle time (", args.MaxIdleTime, ") was exceeded.")
+						log.Println("We proceed to shutdown...")
+
+						Exit()
+
+						return
 					}
 				}
 
-				config.Volumes = append(config.Volumes, outputVolume, submissionVolume)
+				if (resp.StatusCode < 200 || resp.StatusCode >= 300) && resp.StatusCode != 404 {
+					log.Println("Error mqreadservice. No messages returned.")
+					isOk = false
+				}
 
-				result, err := docker.DockerExec(ctx, config)
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					log.Println(err)
+					isOk = false
+				}
 
+				// check payload against json scheme to validate it
+				jsonDocument := gojsonschema.NewBytesLoader(body)
+
+				result, err := schema.Validate(jsonDocument)
 				if err != nil {
-					// There will be some logs already, so just log request json
-					msgJson, errJson := json.Marshal(msg)
-					if errJson != nil {
-						// we are deeply screwed...
-						log.Panicln("<3>Cannot format message as json!", errJson)
-					}
+					log.Println("<3>Cannot validate json against schema", err)
+					isOk = false
+				}
 
-					log.Println("<3>Error while running docker image on message", string(msgJson))
-					return err.Error()
+				if !result.Valid() {
+					log.Println("<3>Json schema validation failed", result.Errors())
+					isOk = false
 				}
 
-				return result
-			}()
+				if err := json.Unmarshal(body, &msg); err != nil {
+					log.Println("<3>Cannot parse json", err)
+					isOk = false
+				}
 
-			// this is path for output from our container (very not standard, but whatever)
-			outputPath := outputVolume.HostPath
+				if isOk {
+					lastGoodRead = time.Now()
+				}
+			}()
 
-			// save whole message there so it will be available in the zip file
-			msgJson, err := json.MarshalIndent(msg, "", " ")
-			if err != nil {
-				log.Println("<4>Cannot save msg to json", err)
-				msgJson = []byte(err.Error())
+			// if there was a problem with getting a TestRun case, then we cannot process it
+			// and we go for next iteration of loop trying to get another TestRun case
+			if !isOk {
+				continue
 			}
 
-			tmp, err := ioutil.TempFile(outputPath, "__msg__*.json")
-			if err == nil {
-				_, err = tmp.Write(msgJson)
-				tmp.Close()
-				defer os.Remove(tmp.Name())
-			}
+			func() {
+				// This is to always log last request, in case something went terribly bad
+				defer func() {
```

Comment:
This should probably go somewhere else now, but that we can decide after everything is finished.

Reply:
ok. leaving it as it is, we sort it out later.